src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs

```diff
@@ -631,189 +610,83 @@ public TypeSymbolWithAnnotations SetUnknownNullabilityForReferenceTypes()
             {
                 Debug.Assert((object)typeSymbol != null);
+                Debug.Assert(!customModifiers.IsDefault);
                 _typeSymbol = typeSymbol;
+                _isNullable = isNullable;
+                _customModifiers = customModifiers;
             }
 
             public sealed override TypeSymbol TypeSymbol => _typeSymbol;
-            public sealed override bool? IsNullable => _typeSymbol.IsNullableType();
-            public override ImmutableArray<CustomModifier> CustomModifiers => ImmutableArray<CustomModifier>.Empty;
+            public sealed override bool? IsNullable => _isNullable;
+            public override ImmutableArray<CustomModifier> CustomModifiers => _customModifiers;
 
             public override TypeSymbolWithAnnotations WithModifiers(ImmutableArray<CustomModifier> customModifiers)
             {
-                if (customModifiers.IsDefaultOrEmpty)
-                {
-                    return this;
-                }
-
-                return new WithCustomModifiers(_typeSymbol, customModifiers);
+                return new NonLazyType(_typeSymbol, _isNullable, customModifiers);
             }
 
             public override TypeSymbol AsTypeSymbolOnly() => _typeSymbol;
 
             // PROTOTYPE(NullableReferenceTypes): Use WithCustomModifiers.Is() => false
             // and set IsNullable=null always for GetTypeParametersAsTypeArguments.
-            public override bool Is(TypeSymbol other) => _typeSymbol.Equals(other, TypeCompareKind.CompareNullableModifiersForReferenceTypes);
+            public override bool Is(TypeSymbol other) => _typeSymbol.Equals(other, TypeCompareKind.CompareNullableModifiersForReferenceTypes) && _customModifiers.IsEmpty;
 
             protected sealed override TypeSymbolWithAnnotations DoUpdate(TypeSymbol typeSymbol, ImmutableArray<CustomModifier> customModifiers)
             {
-                return TypeSymbolWithAnnotations.Create(typeSymbol, customModifiers);
+                return new NonLazyType(typeSymbol, _isNullable, customModifiers);
             }
 
             public override TypeSymbolWithAnnotations AsNullableReferenceType()
             {
-                if (_typeSymbol.IsNullableType())
-                {
-                    return this;
-                }
-
-                return new NullableReferenceTypeWithoutCustomModifiers(_typeSymbol);
+                return _isNullable == true ?
+                    this :
+                    new NonLazyType(_typeSymbol, isNullable: true, _customModifiers);
             }
 
             public override TypeSymbolWithAnnotations AsNotNullableReferenceType()
             {
-                return this;
+                return _isNullable == false ?
+                    this :
+                    new NonLazyType(_typeSymbol, isNullable: false, _customModifiers);
             }
 
             public override string ToDisplayString(SymbolDisplayFormat format)
```

Comment:
Can this implementation be pulled up to the base class? Would it also cover the logic of `LazyNullableType`?

Reply:
Added `// PROTOTYPE` comment.

---
In reply to: [185277255](https://github.com/dotnet/roslyn/pull/26528#discussion_r185277255) [](ancestors = 185277255)